### ISSUE TITLE
Promise.finally is being overwritten in React Native

### ISSUE THREAD

**Issue by benjaminreid** [storybookjs/react-native#20]
**Describe the bug**
The inclusion of Storybook inside a React Native app seems to remove the implementation of `Promise.prototype.finally`. Causing the `Promise.resolve(...).finally is not a function error` to occur.

**To Reproduce**
1. Create a fresh React Native project.
2. Setup Storybook as reference here, https://storybook.js.org/docs/guides/guide-react-native/ `npx -p @storybook/cli sb init --type react_native` (no server)
3. Add the code snippet reference below and you’ll see initially you can call `toString` on `Promise.prototype.finally` but after importing Storybook (you have to use `require` in this instance, not `import`, so the ordering works for the test to see that `finally` starts off being an actual function) it’s no undefined.

Example repo here, https://github.com/benjaminreid/SBPromiseTest

**Expected behavior**
`Promise.prototype.finally` to no be removed.

**Screenshots**
N/A

**Code snippets**
`finally` is valid here, but after the inclusion of Storybook, it’s `undefined`.

```js
console.log('Before:', Promise.prototype.finally.toString());

const StoryBook = require('./storybook');

console.log('After:', Promise.prototype.finally);
```

**System:**
```
Environment Info:

  System:
    OS: macOS Mojave 10.14.6
    CPU: (4) x64 Intel(R) Core(TM) i5-7360U CPU @ 2.30GHz
  Binaries:
    Node: 10.16.0 - /var/folders/94/btf_scw12xbdnqv5dtv2t2hr0000gn/T/fnm-shell-8148567/bin/node
    Yarn: 1.17.3 - /usr/local/bin/yarn
    npm: 6.9.0 - /var/folders/94/btf_scw12xbdnqv5dtv2t2hr0000gn/T/fnm-shell-8148567/bin/npm
  Browsers:
    Chrome: 77.0.3865.90
    Firefox: 69.0.1
    Safari: 13.0.2
  npmPackages:
    @storybook/react-native: ^5.3.0-alpha.17 => 5.3.0-alpha.17
```

**Additional context**
In my instance `redux-persist` use this in their `onBeforeLift` callback.

**Comment by joaocarmo** [comment]
A good set it and forget it workaround for this issue is a combination of the solutions presented by @tjbenton and @limaAniceto above with [patch-package](https://www.npmjs.com/package/patch-package):

1. Open the entry file

```sh
vi node_modules/@storybook/react-native/dist/index.js 
```

2. Append the fixes to the start and end of the file

```js
var fnFix = Promise.prototype.finally;
/*original code*/
Promise.prototype.finally = fnFix;
```

3. Save and commit the patch to your repository

```sh
yarn patch-package @storybook/react-native && git add patches
```

This will ensure you do not include in your project patches to modules, that the package is always patched for every developer in your team and can even be included in CI/CD pipelines and containers.

The diff is saved in a `patches` directory and when the faulty module gets fixed, you can just remove it.

**Comment by colinta** [comment]
This issue just cost me a night and a morning trying to hunt down what was going on, so to find out it's a storybook issue _and it's been known about_ really has steam coming out of my ears.

**Comment by benjaminreid** [comment]
@colinta You’ve just reminded me this issue exists, I was hoping with version 6 this may have been closed. Shame if it’s still hanging around.

**Comment by dannyhw** [comment]
looking into a fix for this, seems like just changing the babel config could resolve this.

__

edit: maybe not as simple as expected still investigating

**Comment by dannyhw** [comment]
This is from corejs and it seems like the problem starts with storybook core because even if I remove all the old polyfills from the @storybook/react-native config then it still happens.

The babel preset env polyfill will cause this when ever 'core-js/modules/es.promise' gets imported, see the following issues:

https://github.com/zloirock/core-js/issues/749
https://github.com/babel/babel/issues/11742

**Comment by apedroferreira** [comment]
@dannyhw any updates? has this been fixed?

**Comment by dannyhw** [comment]
@apedroferreira unfortunately not since the issue is upstream I wasn't able to make progress on this. I'm hoping that maybe upgrading dependencies to 6.0+ will help. I'll test it in the 6.0 upgrade branch. 

From what I understand there would need to be a version of storybook client that isn't polyfilled or at least with a different browser target. So I think a workaround is the best chance we have for now, however I will raise the issue on the storybook repo to see if we can look into it.

Edit: In the past I did also try applying the workaround above to the project and it doesn't seem to work for me

**Comment by ajacquierbret** [comment]
@dannyhw Today, after doing an upgrade of the Expo SDK (40), the same error appeared.

After some investigations, more or less dumb tries of making my app working, I discovered something quite interesting.

If I add `import 'core-js/features/promise/finally'` to `@storybook/react-native/dist/index.js` it still shows an error.
But if I remove the import and add it again after my app is fully loaded, and then save the file, the Fast Refresh feature of react-native handles the import and  then everything works until the next JS bundle reload.

That means that if I want to work on my app, I need to add this import after every bundle reload. And if I want to ship it in production, I need to do the same procedure before doing any build. I desperately need help.

I honestly don't know what to do with this info but maybe someone will.

Btw, the above workaround did work for a while, but only for a while. I'm now quite stuck with this error and don't have any clue on how to get my app working. Any fix or workaround would be HUGELY appreciated.

**Comment by ajacquierbret** [comment]
Ok I managed to get my app working ! 

Adding an import of `core-js/features/promise` AFTER importing Storybook is the fix. Above workarounds and patches don't work because `Promise.prototype.finally` keeps being overwritten even after the execution of `@storybook/react-native/dist/index.js`, maybe because of some Storybook plugins.

In my case, I had to import `core-js/features/promise` a bit deep in my app hierarchy, but this effectively resolves the issue.

Anyway, some consideration about this issue from the Storybook team would be highly appreciated, since any RN developer working with promise based libs (e.g. @apollo/client 3.x) will be facing the same issue...

**Comment by dannyhw** [comment]
@ajacquierbret hey thanks for posting your workaround hopefully that will save some people a lot of frustration. 

I've tried different things in the past to resolve this and the proper fix always seems to require changes upstream. It's hard to coordinate these changes since it's only rn with this problem. I think in the web versions it's fixed via webpack wizardry.

I was kind of hoping that 6.0 might fix this but I'm not sure yet. I'll try it out in the upgrade branch soon.

**Comment by 4lun** [comment]
Posting in case I can help save someone a few hours of their time, it looks like this polyfill behaviour of Storybook breaks `promise.resolve` in [Native Module contexts](https://reactnative.dev/docs/native-modules-android#promises) too.

Example in a [native Android module](https://github.com/4lun/react-native-awesome-module/blob/eb033343f242cccfe604fcdee308227a751d2efe/android/src/main/java/com/reactnativeawesomemodule/AwesomeModuleModule.kt#L16-L21
)
```kotlin
@ReactMethod
fun multiply(a: Int, b: Int, promise: Promise) {
  promise.resolve(a * b)
}
```

Importing any part of Storybook in a project breaks the above example (the promise is never resolved on the RN/JS side), e.g.
```javascript
import {getStorybookUI} from '@storybook/react-native';
getStorybookUI({});

// The following will now never resolve (but did work prior to the inclusion of Storybook)
AwesomeModule.multiply(3,7).then(...); // or
(async () => { const res = await AwesomeModule.multiply(3,7); ... })();
```

[MVP repo with screenshots](https://github.com/4lun/react-native-with-awesome-module-import/commit/b6c287b0c0a1422e4212c0dc2745f227d1f9c38a#diff-a369d17a5dcc6a08663afe8ec644ed0ebc2624f39b704924078c01fd52c58fa3R23), basic example bootstrapped using `react-native init` & `create-react-native-library`

Have tried the potential fixes mentioned in this thread, but currently I've been unsuccessful in figuring a way around the issue.

**Comment by dannyhw** [comment]
@4lun thanks for providing information and examples.

I will try out your example and see what comes up. In my projects I don't have issues with promises so I'm sure there is a way around this.

**Comment by dannyhw** [comment]
@4lun I've experimented a bit and the promise not returning only happens if you run the `getStorybookUI({})` function call from what I can tell.
I'm still investigating so I don't have all the details, nor can I offer a proper solution yet.

Heres some stuff that may or may not be useful 🤷 .
In some cases you can get around this by not calling this function in the top of your app.tsx and instead have a function that will set up the UI when you need it.
like `const getSBUI = () => getStorybookUI({asyncStorage: null});` 
then you can do something like
``` 
 if (showSB) {
    const SBUI = getSBUI();
    return <SBUI />;
  }
```

It's kind of a workaround for the case where you want to have storybook optionally render without it breaking other things in the app. 

I realise it's not exactly the solution to your problem however it does make it possible to include storybook along side other react native code without breaking things when storybook isn't rendered.

Also it does confirm it's not the @react-native/storybook import. I'm still going to investigate further and see if I can find which line in the getStorybookUI function is causing this.

My theory on this is still that the polyfill issue comes from the storybook client api (found in the web storybook repo) which uses polyfills etc. I'm fairly confident in this because I've tried removing all polyfills from react-native storybook and it made no difference.

**Comment by dannyhw** [comment]
@4lun So I actually ended up finding what I think is the cause of the problem.

If you open up
node_modules/@storybook/addons/dist/index.js

and comment out line 11 `//require("core-js/modules/es.promise");` like this then it starts working.

Again not an actual solution yet but from here I can try and figure out if there is a way around this.

I'm going to need help from someone on the storybook team probably because I'm not sure there is any reasonable way to fix this for react native without changes to the pollyfill on storybook addons

@ndelangen or @shilman if you have a moment maybe you can chime in on this? It's been brought up before but I wonder if maybe there is any way to avoid including corejs/es.promise? I looked on the latest version of addons and it still has this import but maybe this will be changing when internet explorer support gets dropped? Seems explorer is the only one without promise support.

**Comment by 4lun** [comment]
Thanks @dannyhw, I've ended up doing something similar to your suggestion, wrapping up all references to Storybook and conditionally loading it via an env flag.

The native module I'm working on shouldn't ever be utilised in a storybook context so this works quite well for the time being.

**Comment by dannyhw** [comment]
A [PR](https://github.com/storybookjs/storybook/pull/14954) was created on the main storybook repository with a fix for future version of storybook. This means that this issue will be resolved in 6.0 of react native storybook. The PR enables provides a modern version of the distributed files, in your project you would add the following config `resolverMainFields:['sbmodern', 'main']` and the storybook dependencies used would be without polyfils.

**Comment by nickplee** [comment]
https://github.com/storybookjs/react-native/issues/159#issuecomment-852422659

**Comment by defrex** [comment]
Since as yet there is no `@storybook/react-native@6`, I've found a more elegant workaround.

The below example lazy-loads Storybook. Convenient both to avoid polyfill downgrades as well as tree-shaking the code out of production builds.

```typescript
import React, { ReactNode, useEffect, useState } from 'react'

export function StorybookScreen() {
  const [storybookUI, setStorybookUI] = useState<ReactNode | null>(null)

  useEffect(() => {
    if (!__DEV__) return
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const { StorybookUI } = require('../../storybook')
    setStorybookUI(<StorybookUI />)
  }, [setStorybookUI])

  return <>{storybookUI}</>
}
```

**Comment by dannyhw** [comment]
This issue should be resolved for good in 6.0 and I have some good news if you're waiting for that.

Now that all the issues in the first milestone have been completed I have released the first alpha.

I've also created an initial guide on how you can get setup with a fresh project. Migration guide to come in the future.

Find the guide here:
https://github.com/storybookjs/react-native/blob/next-6.0/v6README.md

**Comment by mmoio** [comment]
Following the approach of @defrex i've used the lazy approach with lazy and suspense:

StorybookScreen.tsx
```javascript
const StorybookUI = lazy(() => import('../../storybook'));

const StorybookScreen: React.FC = () => {
  return (
    <Suspense fallback={<Text> loading </Text>}>
      <StorybookUI />
    </Suspense>
  );
};
```
App.tsx
```javascript
const App: React.FC = () => (
  { __DEV__? <StorybookScreen/> : <App/> }
);
```

**Comment by mmoio** [comment]
Following the approach of @defrex [comment](https://github.com/storybookjs/react-native/issues/20#issuecomment-861033581) i've been able to get it to work with react lazy and suspense:

StorybookScreen.tsx
```javascript
const StorybookUI = lazy(() => import('../../storybook'));

const StorybookScreen: React.FC = () => {
  return (
    <Suspense fallback={<Text> loading </Text>}>
      <StorybookUI />
    </Suspense>
  );
};
```
App.tsx
```javascript
const App: React.FC = () => (
  { __DEV__? <StorybookScreen/> : <App/> }
);
```

**Comment by dannyhw** [comment]
In this [PR](https://github.com/HurrellT/formikStorybookTest/pull/1) I demonstrate how you can use patch package to fix this error.

Basically just this patch for @storybook/addons

```
diff --git a/node_modules/@storybook/addons/dist/index.js b/node_modules/@storybook/addons/dist/index.js
index 589d611..81bc9e7 100644
--- a/node_modules/@storybook/addons/dist/index.js
+++ b/node_modules/@storybook/addons/dist/index.js
@@ -8,8 +8,6 @@ require("core-js/modules/es.object.to-string");
 
 require("core-js/modules/es.object.values");
 
-require("core-js/modules/es.promise");
-
 require("core-js/modules/web.dom-collections.for-each");
 
 Object.defineProperty(exports, "__esModule", {
 
```

**Comment by pke** [comment]
None of the proposed fixes work. Not even after restarting the metro server. 

Removing the import from @storybook/addons/dist/index.js causes a stack exceed exception on app start.

I also tried not to call `getStorybookUI` and that causes some weird timeout problems in core-js microtasks.

At this state the lib is not usable if it destroys the RN promise implementation.

The only way to get this to work is to use a timeout (of 10 ms) before loading any storybook code.

```ts
function StorybookScreen() {
  const [storybookUI, setStorybookUI] = React.useState<ReactNode | null>(null)

  useEffect(() => {
    if (!__DEV__) {
      return
    }
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const timer = setTimeout(() => {
      const { StorybookUI } = require("../storybook")
      setStorybookUI(<StorybookUI />)
    }, 10)
    return () => clearTimeout(timer)
  }, [setStorybookUI])

  return (
    <>
      {storybookUI}
     </>
  )
}
``

**Comment by dannyhw** [comment]
@pke patching the addons dependency to remove the corejs promise polyfill works as a workaround as far as I know. 

Also there was a new version yesterday to fix some issues in the latest version of react native, not sure if its related to any issues you're having.

**Comment by pke** [comment]
Removing the polyfill didn't work in my setup. I tried for hours. It worked in hot reload sometimes but not deterministic. Cold boot the app never worked. The timeout works. It's clearly a race condition and my setup might be slower or faster than the rest of the ones who tries the workarounds. 

I am on a fairly old RN version 66.3 and I still don't understand how the polyfill can even be active when RN provides a complete es6 promise including finally. I'd like to understand how the polyfill works. 

In the corejs I found related bug and they clearly stated the way storybook authors use the polyfills is wrong. They shouldn't be imported from the modules folder.

**Comment by dannyhw** [comment]
@pke in the v6 beta we get around this problem by resolving to the modern build of storybook (bundled as sbmodern)

**Comment by pke** [comment]
now just checking back with my app its not working again... haha seems to be clock related or something weird.
Anyway... I am going to give v6 a try. However I really would like to use the webserver as a co-pilot, which is not yet ready in v6 iiuc?

**Comment by dannyhw** [comment]
The server is not ready and probably won't be for some time. I don't really know how to make it work for v6 and if I were to recreate it I'd probably opt for implementing something a bit different.

As always if someone is willing to figure it out then I'll happily take a PR but I've tried a few times and it seems to not be compatible without making a lot of changes.

**Comment by pke** [comment]
> @pke patching the addons dependency to remove the corejs promise polyfill works as a workaround as far as I know.

I tried again removing the import and now it corejs runs into a stack overflow cause no promise seems to be defined and it uses some alternative setTimeout path or so.

![image](https://user-images.githubusercontent.com/164625/198990533-74e9fa64-0b92-4675-973a-22007bdfddee.png)

That is on RN 66.3

**Comment by dannyhw** [comment]
@pke if you have a reproduction I can try to debug it for you.

**Comment by pke** [comment]
Thanks @dannyhw, I can't produce a version where its easily reproducable, thats the problem. It seems to be totally random.
What I want to understand is: how does the polyfill thinks it needs to engage? RN has a complete Promise implementation and it should be loaded with RN at boot and therefore available when core-js runs. 
Or has it something to do with RN itself depending on core-js internally, but just a different version than storybook addons?

**Comment by Waltari10** [comment]
Hey, I tried upgrading to Storybook 6 and I still experience this problem with Apollo-client. Removing storybook fixes the issue. Any idea what to do?


    "@storybook/addon-actions": "6.5.13",
    "@storybook/react-native": "6.0.1-beta.10",
    "react-native": "0.70.3",
    "@apollo/client": "3.7.1",

**Comment by dannyhw** [comment]
@Waltari10 make sure you add the metro config to resolve sbmodern.

**Comment by MiguelAraCo** [comment]
An alternative solution that should work for any version is to configure Metro so that it doesn't load the polyfill that Storybook packages load. 

Inside `metro.config.js` add something like this:

```javascript
const { getDefaultConfig } = require('expo/metro-config');

const config = getDefaultConfig(__dirname);

config.resolver.resolveRequest = (context, moduleName, platform) => {
  if (moduleName.startsWith('core-js/modules/es.promise')) {
    // Storybook patches Promise and removes finally, causing errors throughout the application
    // This configures it so that the offending package isn't loaded when requested
    return {
      type: 'empty',
    };
  } else {
    return context.resolveRequest(context, moduleName, platform);
  }
};

module.exports = config;
```

**Comment by tj-mc** [comment]
Is there a more up-to-date solution to this? I'm trying to use `@rnx-kit/metro-resolver-symlinks`, but I must overwrite `resolveRequest`. It seems like this issue is affecting almost everyone using storybook in React Native, and it can be very difficult to find out that storybook is the source of it.

**Comment by tj-mc** [comment]
I have updated to the latest version, as I saw in a changelog that this issue was fixed, however it persists in my case
I have set 

`config.resolver.resolverMainFields = ['sbmodern', 'react-native', 'browser', 'main'];`

In `metro.config.js`

**Comment by dannyhw** [comment]
@tj-mc  what version are you using? Its only fixed in 6.5.

**Comment by tj-mc** [comment]
Turns out I needed to upgrade the other `@storybook` deps too. For anyone solving this issue, here is what I needed. 

```
 "@storybook/addon-actions": "^7.0.9",
 "@storybook/addon-knobs": "^7.0.2",
 "@storybook/addon-links": "^7.0.9",
 "@storybook/addon-ondevice-actions": "^6.5.3",
 "@storybook/addon-ondevice-knobs": "^6.5.3",
 "@storybook/react-native": "^6.5.3",
 "@storybook/react-native-server": "^6.5.3",
 
  "react-native": "0.69.8",
```

In `metro.config.js`:

`config.resolver.resolverMainFields = ['sbmodern', 'react-native', 'browser', 'main'];`

The last 3 values are the default for React Native.

And for the record, this is an ejected Expo app. Now working fine without any errors. Thank you for your work on this! ❤️ 


If you were previously using the [above custom-resolver fix](https://github.com/storybookjs/react-native/issues/20#issuecomment-1345689860), you can now delete it 😄
